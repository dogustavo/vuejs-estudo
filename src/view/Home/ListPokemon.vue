<script setup lang="ts">
import { ref, onMounted, watch } from 'vue'
import { useRoute } from 'vue-router'

import Container from '@/components/Container/AppContainer.vue'
import PokemonCard from '@/components/PokemonCard/PokemonCard.vue'

import { getPokemons } from '@/services'
import type { Pokemon } from '@/types/pokemon'

import styles from './styles.module.scss'
import PaginationComponent from '@/components/Paginate/PaginationComponent.vue'

const route = useRoute()

const isLoading = ref(true)
const error = ref('')
const pokemonsData = ref<Pokemon[]>()
const pageInfo = ref<{
  count: number
  limit: number
  offset: number
  next: null | string
  previous: null | string
}>({
  count: 0,
  limit: 21,
  offset: 0,
  next: null,
  previous: null,
})

const searchTerm = ref(route.query)

const loadPokemon = async () => {
  try {
    const pokemons = await getPokemons({
      filter: {
        limit: searchTerm.value.limit || 21,
        offset: searchTerm.value.offset || 0,
      },
    })

    pokemonsData.value = pokemons.data
    pageInfo.value = {
      count: pokemons.count ?? 1,
      offset: +(searchTerm.value.offset ?? 0),
      limit: +(searchTerm.value.limit ?? 21),
      next: pokemons.next || null,
      previous: pokemons.previous || null,
    }
  } catch (err: unknown) {
    if (err instanceof Error) {
      error.value = err.message
    }
  } finally {
    isLoading.value = false
  }
}

onMounted(loadPokemon)

watch(
  () => route.query,
  async (newQuery) => {
    searchTerm.value = newQuery

    isLoading.value = true
    await loadPokemon()
  },
  { immediate: true },
)
</script>

<template>
  <section :class="styles['pokemons-section']">
    <Container>
      <div :class="styles.list">
        <router-link
          v-for="pokemon in pokemonsData"
          v-bind:key="pokemon.id"
          :to="`/${pokemon.name}`"
        >
          <PokemonCard :pokemon="pokemon" />
        </router-link>
      </div>

      <PaginationComponent
        :limit="pageInfo.limit"
        :offset="pageInfo.offset"
        :count="pageInfo.count"
        :next="pageInfo.next"
        :previous="pageInfo.previous"
      />
    </Container>
  </section>
</template>

<style scoped></style>
