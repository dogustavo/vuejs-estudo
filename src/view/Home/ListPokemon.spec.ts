import { describe } from 'node:test'
import { test, expect, vi } from 'vitest'

import ListPokemon from './ListPokemon.vue'
import { shallowMount } from '@vue/test-utils'

vi.mock('vue-router', () => ({
  useRoute: () => ({
    query: {
      limit: 21,
      offset: 0,
    },
  }),
  useRouter: () => ({
    push: vi.fn(),
    // Mock de outras funcionalidades do `useRouter`, se necessÃ¡rio
  }),
}))

vi.mock('@/services', () => ({
  getPokemons: vi.fn().mockResolvedValue({
    data: [
      {
        id: 1,
        name: 'bulbasaur',
        sprites: {
          front_default:
            'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png',
        },
        stats: [],
        types: [],
      },
    ],
    count: 2,
    next: null,
    previous: null,
  }),
}))

const renderComponent = () => shallowMount(ListPokemon)

describe('ListPokemon.vue', async () => {
  test('Renderiza o componente de forma correta', () => {
    const wrapper = renderComponent()

    expect(wrapper.exists()).toBe(true)
    expect(wrapper.html()).toMatchSnapshot()
  })
})
