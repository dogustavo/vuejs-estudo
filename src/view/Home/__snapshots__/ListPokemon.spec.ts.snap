// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Renderiza o componente de forma correta 1`] = `
"<section data-v-5e3cb3a3="" class="pokemons-section">
  <container-stub data-v-5e3cb3a3=""></container-stub>
</section>"
`;
